layer/profiles.cpp

```diff
@@ -4056,6 +4290,248 @@ bool JsonLoader::valid(ExtensionSupport support) {
+    }
+
+    return valid;
+}
+
+bool QueueFamilyAndExtensionsMatch(const QueueFamilyProperties &device, const QueueFamilyProperties &profile) {
+    if (!QueueFamilyMatch(device.properties_2.queueFamilyProperties, profile.properties_2.queueFamilyProperties)) {
+        return false;
+    }
+    if (!GlobalPriorityMatch(device.global_priority_properties_, profile.global_priority_properties_)) {
+        return false;
+    }
+    if ((device.video_properties_.videoCodecOperations & profile.video_properties_.videoCodecOperations) !=
+        profile.video_properties_.videoCodecOperations) {
+        return false;
+    }
+    if ((device.checkpoint_properties_.checkpointExecutionStageMask &
+         profile.checkpoint_properties_.checkpointExecutionStageMask) !=
+        profile.checkpoint_properties_.checkpointExecutionStageMask) {
+        return false;
+    }
+    if ((device.checkpoint_properties_2_.checkpointExecutionStageMask &
+         profile.checkpoint_properties_2_.checkpointExecutionStageMask) !=
+        profile.checkpoint_properties_2_.checkpointExecutionStageMask) {
+        return false;
+    }
+    if (device.query_result_status_properties_.supported != profile.query_result_status_properties_.supported) {
+        return false;
+    }
+    return true;
+}
+
+bool JsonLoader::OrderQueueFamilyProperties(ArrayOfVkQueueFamilyProperties* qfp) {
+    if (qfp->empty()) {
+        return true;
+    }
+    // If device has less queue families than needed we can't load all profile queue families
+    if (pdd_.device_queue_family_properties_.size() < qfp->size()) {
+        return false;
+    }
+    // Find all permutations and see if any of them supports profile queue families
+    std::vector<uint32_t> permutations(pdd_.device_queue_family_properties_.size());
+    uint32_t count = static_cast<uint32_t>(permutations.size());
+    for (uint32_t i = 0; i < count; ++i) {
+        permutations[i] = i;
+    }
+    do {
+        bool match = true;
+        for (uint32_t i = 0; i < count; ++i) {
+            if (permutations[i] < qfp->size() && !QueueFamilyAndExtensionsMatch(pdd_.device_queue_family_properties_[i], (*qfp)[permutations[i]])) {
+                match = false;
+                break;
+            }
+        }
+        if (match) {
+            // Empty queue families at the end are not needed
+            while (permutations[count - 1] >= qfp->size()) {
+                --count;
+            }
+            ArrayOfVkQueueFamilyProperties ordered;
+            for (uint32_t i = 0; i < count; ++i) {
+                if (permutations[i] < qfp->size()) {
+                    ordered.push_back((*qfp)[permutations[i]]);
+                } else {
+                    ordered.push_back(QueueFamilyProperties());
+                }
+            }
+            *qfp = ordered;
+            return true;
+        }
+    } while (std::next_permutation(permutations.begin(), permutations.end()));
```

Comment:
This checks that all requested queues can actually be used and allows the layer to reorder them so that the profile queue will be on the same index that the device queue that supports it is.

Reply:
Yes, I got that. I just don't think you have to test for every permutation, which requires `O(n * n!)` time complexity, as it should be doable in `O(n^2)`, I believe.

Reply:
Though probably it doesn't matter as `n` won't ever be really large.

Reply:
I thought about time complexity too, but this code will be hard to test, so I went with the safe version for now.

Reply:
Yes, I concur. I've tweaked the library code to work similarly (see PR #129). If you could review that, I'd appreciate it.